### PR TITLE
Homework#2 completed

### PR DESCRIPTION
**Pull request by Wizzard008**

### CODE REVIEW

File: src/test/java/ru/training/at/hw2/ex2/WebPageTestsCommonConditions.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package ru.training.at.hw2.ex2;
+
+import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+
+public class WebPageTestsCommonConditions {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/WebPageTestsCommonConditions.java]: looks like exact copy of `ru.training.at.hw2.ex1.WebPageTestsCommonConditions`
Please fix unnecessary duplication

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        String expectedTitle = "Home Page";
+        String actualTitle = driver.getTitle();
+        Assert.assertEquals(actualTitle, expectedTitle);
+
+        //3.Perform login
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .elementToBeClickable(By.className("profile-photo"))).click();
+
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .elementToBeClickable(By.id("name"))).sendKeys(LOGIN_NAME);
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Linebreaks are hard to read. Compare to the following
```suggestion
        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
            .until(ExpectedConditions.elementToBeClickable(By.id("name")))
            .sendKeys(LOGIN_NAME);
```

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            .elementToBeClickable(By.id("password"))).sendKeys(PASSWORD);
+
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .elementToBeClickable(By.id("login-button"))).click();
+
+        //4. Assert Username is loggined
+        String actualDisplayedUsername = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .presenceOfElementLocated(By.id("user-name"))).getText();
+
+        String expectedUsername = "ROMAN IOVLEV";
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java:48]: Input data (login/password) and expected data is stored in different places, but all of them are "test input data" and should be treated in the same way

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        Assert.assertEquals(actualDisplayedUsername, expectedUsername);
+
+        //5. Assert that there are 4 items on the header section are displayed and they have proper texts
+        List<WebElement> headerMenuButtons = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .presenceOfAllElementsLocatedBy(By.xpath("//div[1]/nav[1]/ul[1]/li/a")));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Please improve locator (remember the order: DOM, CSS, XPATH)

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+        Assert.assertEquals(headerMenuButtons.size(), 4);
+        Assert.assertEquals(headerMenuButtons.get(0).getText(), "HOME");
+        Assert.assertEquals(headerMenuButtons.get(1).getText(), "CONTACT FORM");
+        Assert.assertEquals(headerMenuButtons.get(2).getText(), "SERVICE");
+        Assert.assertEquals(headerMenuButtons.get(3).getText(), "METALS & COLORS");
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: It looks like one assertion

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+        //6. Assert that there are 4 images on the Index Page and they are displayed
+        List<WebElement> iconsImages = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .presenceOfAllElementsLocatedBy(By.xpath("//div[contains(@class,\"icon\")]")));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Please improve locator (CSS is better choice here)

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        String expectedThirdBoxText = "To be multiplatform";
+        String expectedForthBoxText = "Already have good base\n"
+            + "(about 20 internal and\n"
+            + "some external projects),\n"
+            + "wish to get more…";
+
+        Assert.assertTrue(iconsTextsWebElements.get(0).getText().equalsIgnoreCase(expectedFirstBoxText));
+        Assert.assertTrue(iconsTextsWebElements.get(1).getText().equalsIgnoreCase(expectedSecondBoxText));
+        Assert.assertTrue(iconsTextsWebElements.get(2).getText().equalsIgnoreCase(expectedThirdBoxText));
+        Assert.assertTrue(iconsTextsWebElements.get(3).getText().equalsIgnoreCase(expectedForthBoxText));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Again, looks like one assert

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        //8.Assert that there is the iframe with “Frame Button” exist
+        Assert.assertTrue(driver.getPageSource().contains("iframe"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: `.getPageSource()` returns a string with Html code, which is not what you really want

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+        //9.Switch to the iframe and check that there is “Frame Button” in the iframe
+        WebElement iframe = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .presenceOfElementLocated(By.id("frame")));
+
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .frameToBeAvailableAndSwitchToIt(iframe));
+
+        Assert.assertTrue(driver.getPageSource().contains("frame-button"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: The same
please look into docs or examples, e.g. [this](https://www.guru99.com/handling-iframes-selenium.html)

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+        //10.Switch to original window back
+        driver.switchTo().defaultContent();
+
+        //11.Assert that there are 5 items in the Left Section are displayed and they have proper text
+        List<WebElement> leftSectionMenuWebElements =
+            new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+                .presenceOfAllElementsLocatedBy(By.xpath("//ul[@class=\"sidebar-menu left\"]/li/a/span")));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Please use more readable selector

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        String expectedMenuItem2 = "Contact form";
+        String expectedMenuItem3 = "Service";
+        String expectedMenuItem4 = "Metals & Colors";
+        String expectedMenuItem5 = "Elements packs";
+        Assert.assertEquals(leftSectionMenuWebElements.get(0).getText(), expectedMenuItem1);
+        Assert.assertEquals(leftSectionMenuWebElements.get(1).getText(), expectedMenuItem2);
+        Assert.assertEquals(leftSectionMenuWebElements.get(2).getText(), expectedMenuItem3);
+        Assert.assertEquals(leftSectionMenuWebElements.get(3).getText(), expectedMenuItem4);
+        Assert.assertEquals(leftSectionMenuWebElements.get(4).getText(), expectedMenuItem5);
+    }
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: This looks like one assert too

File: src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        //4. Assert Username is loggined
+        String actualDisplayedUsername = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .presenceOfElementLocated(By.id("user-name"))).getText();
+
+        String expectedUsername = "ROMAN IOVLEV";
+        Assert.assertEquals(actualDisplayedUsername, expectedUsername);
+
+        //5.Open through the header menu Service -> Different Elements Page
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .elementToBeClickable(By.xpath("//li[@class=\"dropdown\"]/*[@class=\"dropdown-toggle\"]"))).click();
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java]: Please use CSS equivalent

File: src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .elementToBeClickable(By.xpath("//a[contains(@href,\"different-elements.html\")]"))).click();
+
+        //6.Select checkboxes - Water, Wind
+        List<WebElement> checkboxMenuItems = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
+            .presenceOfAllElementsLocatedBy(By.className("label-checkbox")));
+        for (WebElement el : checkboxMenuItems
+        ) {
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java]: Please use more Javaic breaks :)
```suggestion
        for (WebElement el : checkboxMenuItems) {
```

File: src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            .elementToBeClickable(By.xpath("//option[text()=\"Yellow\"]"))).click();
+
+        //9.Assert that
+        //• for each checkbox there is an individual log row and value is corresponded to the status of checkbox
+        //• for radio button there is a log row and value is corresponded to the status of radio button
+        //• for dropdown there is a log row and value is corresponded to the selected value
+        Assert.assertTrue(driver.getPageSource().contains("Colors: value changed to Yellow"));
+        Assert.assertTrue(driver.getPageSource().contains("metal: value changed to  Selen"));
+        Assert.assertTrue(driver.getPageSource().contains("Wind: condition changed to true"));
+        Assert.assertTrue(driver.getPageSource().contains("Water: condition changed to true"));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java]: replace `.getPageSource()` with other approach

**Review by shoggy — CHANGES_REQUESTED** [review]

File: src/test/java/ru/training/at/hw2/ex2/WebPageTestsCommonConditions.java (hunk 1 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex2/WebPageTestsCommonConditions.java]: WebPageTestsCommonConditions moved to util package.

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 3 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java:48]: Corrected

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 4 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java:45]: Corrected: By.xpath("//*[@class=\"uui-navigation nav navbar-nav m-l8\"]/li")

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 5 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected.  
        String[] actualHeaderMenuTitles = headerMenuButtons.stream().map(WebElement::getText).toArray(String[]::new);
        String[] expectedHeaderMenuTitles = {"HOME","CONTACT FORM","SERVICE","METALS & COLORS"};
        Assert.assertEquals(actualHeaderMenuTitles,expectedHeaderMenuTitles);

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 6 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Changed to By.cssSelector("[class=\"benefit-icon\"]"

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 3 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java:48]: String userLogin = "Roman";
        String userPassword = "Jdi1234";
        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
            .until(ExpectedConditions.elementToBeClickable(By.id("name")))
            .sendKeys(userLogin);

        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
            .until(ExpectedConditions.elementToBeClickable(By.id("password")))
            .sendKeys(userPassword);

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 4 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected
By.cssSelector("[class=\"uui-navigation nav navbar-nav m-l8\"] > li"

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 7 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected
        String[] actualIconsTexts = iconsTextsWebElements.stream().map(WebElement::getText).toArray(String[]::new);
        String[] expectedIconsTexts = {expectedFirstBoxText,expectedSecondBoxText,expectedThirdBoxText,expectedForthBoxText};
        Assert.assertEquals(actualIconsTexts,expectedIconsTexts);

File: src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java (hunk 13 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java]: Corrected

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 8 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected
        WebElement iframe = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
            .until(ExpectedConditions.presenceOfElementLocated(By.id("frame")));
        Assert.assertTrue(iframe.isDisplayed());

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 9 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected
        WebElement frameButtonWebElement = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
            .until(ExpectedConditions.presenceOfElementLocated(By.id("frame-button")));

        Assert.assertTrue(frameButtonWebElement.isDisplayed());

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 10 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Changed
By.cssSelector("#mCSB_1_container > ul > li > a > span"

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 11 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected
        String[] actualMenuItemsTexts =
            leftSectionMenuWebElements.stream().map(WebElement::getText).toArray(String[]::new);
        String[] expectedMenuItemsTexts = {"Home", "Contact form", "Service", "Metals & Colors", "Elements packs"};
        Assert.assertEquals(actualMenuItemsTexts, expectedMenuItemsTexts);

File: src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java (hunk 12 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java]: Changed into
By.cssSelector("[class=\"dropdown-toggle\"]")

File: src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java (hunk 14 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex2/EpamDemoWebPageTests.java]: Corrected with softAssert
        String[] actualLogItemsTexts = webpageLogItems.stream().map(WebElement::getText).toArray(String[]::new);
        String[] expectedPartialLogItemsTexts =
            {"Colors: value changed to Yellow", "metal: value changed to  Selen", "Wind: condition changed to true",
                "Water: condition changed to true"};

        SoftAssert softAssert = new SoftAssert();
        softAssert.assertEquals(actualLogItemsTexts,expectedPartialLogItemsTexts);

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 2 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 15]
```diff
@@ -21,46 +22,51 @@ public void demoPageTests() {
...
+                .presenceOfAllElementsLocatedBy(By.cssSelector("[class=\"uui-navigation nav navbar-nav m-l8\"] > li")));
+        String[] actualHeaderMenuTitles = headerMenuButtons.stream().map(WebElement::getText).toArray(String[]::new);
+        String[] expectedHeaderMenuTitles = {"HOME", "CONTACT FORM", "SERVICE", "METALS & COLORS"};
+        Assert.assertEquals(actualHeaderMenuTitles, expectedHeaderMenuTitles);
 
         //6. Assert that there are 4 images on the Index Page and they are displayed
-        List<WebElement> iconsImages = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
-            .presenceOfAllElementsLocatedBy(By.xpath("//div[contains(@class,\"icon\")]")));
+        List<WebElement> iconsImages = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
+            .until(ExpectedConditions.presenceOfAllElementsLocatedBy(By.cssSelector("[class=\"benefit-icon\"]")));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]:
```suggestion
            .until(ExpectedConditions.presenceOfAllElementsLocatedBy(By.cssSelector(".benefit-icon")));
```

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java [hunk 16]
```diff
@@ -74,39 +80,37 @@ public void demoPageTests() {
...
         //9.Switch to the iframe and check that there is “Frame Button” in the iframe
-        WebElement iframe = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
-            .presenceOfElementLocated(By.id("frame")));
+        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
+            .until(ExpectedConditions.frameToBeAvailableAndSwitchToIt(iframe));
 
-        new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS).until(ExpectedConditions
-            .frameToBeAvailableAndSwitchToIt(iframe));
+        WebElement frameButtonWebElement = new WebDriverWait(driver, WAIT_TIMEOUT_SECONDS)
+            .until(ExpectedConditions.presenceOfElementLocated(By.id("frame-button")));
```
Comment (shoggy) [inline src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Somewhere here you need to switch into iframe

**Review by shoggy — CHANGES_REQUESTED** [review]

File: src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java (hunk 16 above)
Comment (Wizzard008) [reply src/test/java/ru/training/at/hw2/ex1/EpamDemoWebPageTests.java]: Corrected

**Review by shoggy — APPROVED** [review]

**Pull request merged**